File: lib/minimal_tebd/validate.py
"""Validate TEBD against exact diagonalisation on small systems.

Usage
-----

    from minimal_tebd import TFIChain, validate_tebd_vs_ed
    model = TFIChain(L=10, J=1.0, g=1.0)
    validate_tebd_vs_ed(model, dt=0.05, t_max=2.0, chi_max=64, order=4)

The function runs the *same* initial state through both TEBD and ED, measures
a small set of observables at each time step, and reports the maximum
elementwise error. Small deviations are expected from Trotter splitting and
finite bond-dimension truncation; the errors should scale as ``dt**order``
for ``chi_max`` large enough that truncation is negligible.
"""

import numpy as np

from .exact_diag import ExactEvolution
from .mps import SimpleMPS
from .tebd import TEBDEngine


def validate_tebd_vs_ed(
    model,
    *,
    dt=0.05,
    t_max=1.0,
    chi_max=64,
    svd_min=1e-12,
    order=4,
    initial_states=None,
    verbose=True,
):
    """Compare TEBD and exact-diag evolution of the same initial product state.

    Parameters
    ----------
    model : TFIChain or BondModel
        ``model.L`` must be small enough that ``d**L`` fits in memory
        (L <= ~14 for spin-1/2).
    dt : float
    t_max : float
    chi_max, svd_min : truncation parameters for TEBD.
    order : int, Trotter order (1, 2, 4).
    initial_states : list[int] or None
        Per-site basis index. Default ``[0] * L`` (``|up ... up>``).
    verbose : bool
        If True, prints per-step errors.

    Returns
    -------
    summary : dict
        ``t``: list of times.
        ``err_Sz``, ``err_Sx``, ``err_corr_XX``, ``err_entropy``:
            arrays of max-absolute-error vs time.
        ``trunc_err``: final accumulated truncation error from TEBD.
        ``fidelity_drop``: ``1 - |<psi_TEBD | psi_ED>|`` at the final time.
    """
    L = model.L
    if L > 14:
        raise ValueError(f'validation needs d**L to fit in memory; L={L} is too large')

    if initial_states is None:
        initial_states = [0] * L

    # TEBD
    psi_mps = SimpleMPS.from_product_state(L, initial_states, d=model.d)
    eng = TEBDEngine(
        psi_mps,
        model,
        {
            'dt': dt,
            'order': order,
            'N_steps': 1,
            'trunc_params': {'chi_max': chi_max, 'svd_min': svd_min},
        },
    )

    # ED
    ed = ExactEvolution.from_product_state(model, initial_states, dt=dt)

    ts = [0.0]
    err_Sz, err_Sx, err_XX, err_S = [], [], [], []

    def compare_step():
        Sz_m = eng.psi.expectation_value('Sigmaz')
        Sz_e = ed.expectation_value('Sigmaz')
        Sx_m = eng.psi.expectation_value('Sigmax')
        Sx_e = ed.expectation_value('Sigmax')
        C_m = eng.psi.correlation_function('Sigmax', 'Sigmax')
        C_e = ed.correlation_function('Sigmax', 'Sigmax')
        S_m = eng.psi.entanglement_entropy()
        S_e = ed.entanglement_entropy()
        err_Sz.append(np.max(np.abs(Sz_m - Sz_e)))
        err_Sx.append(np.max(np.abs(Sx_m - Sx_e)))
        err_XX.append(np.max(np.abs(C_m - C_e)))
        err_S.append(np.max(np.abs(S_m - S_e)))

    compare_step()
    if verbose:
        print(
            f'{"t":>6s}  {"max|dSz|":>10s}  {"max|dSx|":>10s}  '
            f'{"max|dXX|":>10s}  {"max|dS|":>10s}  {"chi":>4s}'
        )
        print(f'{0.0:>6.3f}  {err_Sz[-1]:>10.2e}  {err_Sx[-1]:>10.2e}  '
              f'{err_XX[-1]:>10.2e}  {err_S[-1]:>10.2e}  {max(eng.psi.bond_dimensions()):>4d}')

    n_steps = int(round(t_max / dt))
    for _ in range(n_steps):
        eng.run()
        ed.step()
        ts.append(eng.evolved_time)
        compare_step()
        if verbose:
            print(f'{ts[-1]:>6.3f}  {err_Sz[-1]:>10.2e}  {err_Sx[-1]:>10.2e}  '
                  f'{err_XX[-1]:>10.2e}  {err_S[-1]:>10.2e}  '
                  f'{max(eng.psi.bond_dimensions()):>4d}')

    # Final global fidelity as a single-number sanity check.
    psi_full_mps = eng.psi.to_state_vector()
    fidelity = abs(np.vdot(psi_full_mps, ed.psi))

    summary = {
        't': np.array(ts),
        'err_Sz': np.array(err_Sz),
        'err_Sx': np.array(err_Sx),
        'err_corr_XX': np.array(err_XX),
        'err_entropy': np.array(err_S),
        'trunc_err': eng.trunc_err.eps,
        'fidelity_drop': 1.0 - fidelity,
    }
    if verbose:
        print()
        print(f'final truncation error (accumulated): {eng.trunc_err.eps:.3e}')
        print(f'final |<psi_TEBD | psi_ED>|         : {fidelity:.10f}')
        print(f'final fidelity drop                  : {summary["fidelity_drop"]:.3e}')
    return summary


if __name__ == '__main__':
    # Quick smoke test when the module is run directly.
    from .model import TFIChain
    model = TFIChain(L=8, J=1.0, g=1.0)
    validate_tebd_vs_ed(model, dt=0.05, t_max=1.0, chi_max=64, order=4)

File: src/core/validate.cpp
#include "validate.h"

#include <algorithm>
#include <cmath>
#include <cstring>
#include <iomanip>
#include <iostream>
#include <vector>

#include "exact_diag.h"
#include "mps.h"
#include "tebd.h"

ValidationSummary validate_tebd_vs_ed(const BondModel& model, double dt,
                                      double t_max, int chi_max,
                                      double svd_min, int order,
                                      const int* initial_states,
                                      bool verbose) {
  int L = model.L, d = model.d;
  assert(L <= 14);

  std::vector<int> init(L, 0);
  if (initial_states) {
    for (int i = 0; i < L; ++i) init[i] = initial_states[i];
  }

  // TEBD setup.
  MPS psi_mps = MPS::product_state(L, init.data(), d, chi_max);
  TEBDEngine eng(&psi_mps, const_cast<BondModel*>(&model), dt, order, 1,
                 chi_max, svd_min);

  // ED setup.
  ExactEvolution ed =
      ExactEvolution::from_product_state(model, init.data(), dt);

  ValidationSummary summary;
  summary.t.push_back(0.0);

  std::vector<double> Sz_m(L), Sz_e(L), Sx_m(L), Sx_e(L);
  std::vector<double> C_m(L * L), C_e(L * L);
  std::vector<double> S_m(L - 1), S_e(L - 1);

  auto compare_step = [&]() {
    psi_mps.expect(PAULI_Z, Sz_m.data());
    ed.expectation_value(PAULI_Z, Sz_e.data());
    psi_mps.expect(PAULI_X, Sx_m.data());
    ed.expectation_value(PAULI_X, Sx_e.data());
    psi_mps.corr(PAULI_X, PAULI_X, C_m.data());
    ed.correlation_function(PAULI_X, PAULI_X, C_e.data());
    psi_mps.entropy(S_m.data());
    ed.entanglement_entropy(S_e.data());

    double e_Sz = 0, e_Sx = 0, e_XX = 0, e_S = 0;
    for (int i = 0; i < L; ++i) {
      e_Sz = std::max(e_Sz, std::abs(Sz_m[i] - Sz_e[i]));
      e_Sx = std::max(e_Sx, std::abs(Sx_m[i] - Sx_e[i]));
    }
    for (int i = 0; i < L * L; ++i)
      e_XX = std::max(e_XX, std::abs(C_m[i] - C_e[i]));
    for (int i = 0; i < L - 1; ++i)
      e_S = std::max(e_S, std::abs(S_m[i] - S_e[i]));

    summary.err_Sz.push_back(e_Sz);
    summary.err_Sx.push_back(e_Sx);
    summary.err_corr_XX.push_back(e_XX);
    summary.err_entropy.push_back(e_S);
  };

  compare_step();

  if (verbose) {
    std::cout << std::setw(6) << "t" << "  " << std::setw(10) << "max|dSz|"
              << "  " << std::setw(10) << "max|dSx|" << "  " << std::setw(10)
              << "max|dXX|" << "  " << std::setw(10) << "max|dS|" << "  "
              << std::setw(4) << "chi" << std::endl;
    std::cout << std::fixed << std::setprecision(3) << std::setw(6) << 0.0
              << "  " << std::scientific << std::setprecision(2)
              << std::setw(10) << summary.err_Sz.back() << "  "
              << std::setw(10) << summary.err_Sx.back() << "  "
              << std::setw(10) << summary.err_corr_XX.back() << "  "
              << std::setw(10) << summary.err_entropy.back() << "  "
              << std::setw(4) << psi_mps.max_bond_dim() << std::endl;
  }

  int n_steps = static_cast<int>(std::round(t_max / dt));
  for (int step = 0; step < n_steps; ++step) {
    eng.run();
    ed.step();
    summary.t.push_back(eng.evolved_time);
    compare_step();

    if (verbose) {
      std::cout << std::fixed << std::setprecision(3) << std::setw(6)
                << eng.evolved_time << "  " << std::scientific
                << std::setprecision(2) << std::setw(10)
                << summary.err_Sz.back() << "  " << std::setw(10)
                << summary.err_Sx.back() << "  " << std::setw(10)
                << summary.err_corr_XX.back() << "  " << std::setw(10)
                << summary.err_entropy.back() << "  " << std::setw(4)
                << psi_mps.max_bond_dim() << std::endl;
    }
  }

  // Final fidelity check.
  Tensor psi_full = psi_mps.to_state_vector();
  int dim = psi_full.size;
  Cdouble overlap = 0.0;
  for (int i = 0; i < dim; ++i)
    overlap += std::conj(psi_full(i)) * ed.psi[i];
  double fidelity = std::abs(overlap);

  summary.trunc_err = eng.trunc_err_eps;
  summary.fidelity_drop = 1.0 - fidelity;

  if (verbose) {
    std::cout << std::endl;
    std::cout << "final truncation error (accumulated): " << std::scientific
              << std::setprecision(3) << summary.trunc_err << std::endl;
    std::cout << "final |<psi_TEBD | psi_ED>|         : " << std::fixed
              << std::setprecision(10) << fidelity << std::endl;
    std::cout << "final fidelity drop                  : " << std::scientific
              << std::setprecision(3) << summary.fidelity_drop << std::endl;
  }

  return summary;
}
